com.hazelcast.jet.core.metrics.MetricsTest issues

_master_ (commit 418a40701628c4fafaa063342de6b9c1adb1be13)

Failed on Oracle JDK 8: http://jenkins.hazelcast.com/job/jet-oss-master-oracle-jdk8/281/testReport/com.hazelcast.jet.core.metrics/

Following tests failed:
- `com.hazelcast.jet.core.metrics.MetricsTest.availableDuringJobExecution`
- `com.hazelcast.jet.core.metrics.MetricsTest.metricsAreJobIndependent`
- `com.hazelcast.jet.core.metrics.MetricsTest.test_sourceSinkTag`

Stacktrace for `com.hazelcast.jet.core.metrics.MetricsTest.availableDuringJobExecution` (and the same for `com.hazelcast.jet.core.metrics.MetricsTest.metricsAreJobIndependent` just from line `com.hazelcast.jet.core.metrics.MetricsTest.lambda$metricsAreJobIndependent$8(MetricsTest.java:392)`):
```
java.lang.AssertionError: Expected measurements for metric 'total', but there were none
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertFalse(Assert.java:64)
	at com.hazelcast.jet.core.metrics.JobMetricsChecker.assertAggregatedMetricValue(JobMetricsChecker.java:75)
	at com.hazelcast.jet.core.metrics.JobMetricsChecker.assertSummedMetricValue(JobMetricsChecker.java:53)
	at com.hazelcast.jet.core.metrics.MetricsTest.lambda$availableDuringJobExecution$4(MetricsTest.java:256)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1356)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1458)
	at com.hazelcast.jet.core.metrics.MetricsTest.availableDuringJobExecution(MetricsTest.java:256)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

Stacktrace for `com.hazelcast.jet.core.metrics.MetricsTest.test_sourceSinkTag`:
```
java.lang.AssertionError: Values should be different. Actual: 0
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failEquals(Assert.java:185)
	at org.junit.Assert.assertNotEquals(Assert.java:199)
	at org.junit.Assert.assertNotEquals(Assert.java:211)
	at com.hazelcast.jet.core.metrics.MetricsTest.lambda$test_sourceSinkTag$11(MetricsTest.java:443)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1356)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1458)
	at com.hazelcast.jet.core.metrics.MetricsTest.test_sourceSinkTag(MetricsTest.java:443)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

Standard output for `com.hazelcast.jet.core.metrics.MetricsTest.availableDuringJobExecution`:
```
Hiccups measured while running test 'availableDuringJobExecution(com.hazelcast.jet.core.metrics.MetricsTest):'
13:32:35, accumulated pauses: 311 ms, max pause: 281 ms, pauses over 1000 ms: 0
13:32:40, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:32:45, accumulated pauses: 65 ms, max pause: 3 ms, pauses over 1000 ms: 0
13:32:50, accumulated pauses: 57 ms, max pause: 4 ms, pauses over 1000 ms: 0
13:32:55, accumulated pauses: 26 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:33:00, accumulated pauses: 31 ms, max pause: 1 ms, pauses over 1000 ms: 0
13:33:05, accumulated pauses: 161 ms, max pause: 5 ms, pauses over 1000 ms: 0
13:33:10, accumulated pauses: 120 ms, max pause: 3 ms, pauses over 1000 ms: 0
13:33:15, accumulated pauses: 28 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:33:20, accumulated pauses: 29 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:33:25, accumulated pauses: 28 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:33:30, accumulated pauses: 30 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:33:35, accumulated pauses: 34 ms, max pause: 4 ms, pauses over 1000 ms: 0
13:33:40, accumulated pauses: 29 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:33:45, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:33:50, accumulated pauses: 28 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:33:55, accumulated pauses: 41 ms, max pause: 1 ms, pauses over 1000 ms: 0
13:34:00, accumulated pauses: 42 ms, max pause: 5 ms, pauses over 1000 ms: 0
13:34:05, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:34:10, accumulated pauses: 30 ms, max pause: 2 ms, pauses over 1000 ms: 0
13:34:15, accumulated pauses: 32 ms, max pause: 2 ms, pauses over 1000 ms: 0
13:34:20, accumulated pauses: 55 ms, max pause: 21 ms, pauses over 1000 ms: 0
13:34:25, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:34:30, accumulated pauses: 35 ms, max pause: 2 ms, pauses over 1000 ms: 0
13:34:35, accumulated pauses: 312 ms, max pause: 293 ms, pauses over 1000 ms: 0
```

Standard output for `com.hazelcast.jet.core.metrics.MetricsTest.metricsAreJobIndependent`:
```
Hiccups measured while running test 'metricsAreJobIndependent(com.hazelcast.jet.core.metrics.MetricsTest):'
13:34:35, accumulated pauses: 323 ms, max pause: 293 ms, pauses over 1000 ms: 0
13:34:40, accumulated pauses: 28 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:34:45, accumulated pauses: 45 ms, max pause: 3 ms, pauses over 1000 ms: 0
13:34:50, accumulated pauses: 32 ms, max pause: 1 ms, pauses over 1000 ms: 0
13:34:55, accumulated pauses: 35 ms, max pause: 4 ms, pauses over 1000 ms: 0
13:35:00, accumulated pauses: 28 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:35:05, accumulated pauses: 32 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:35:10, accumulated pauses: 37 ms, max pause: 1 ms, pauses over 1000 ms: 0
13:35:15, accumulated pauses: 31 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:35:20, accumulated pauses: 29 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:35:25, accumulated pauses: 29 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:35:30, accumulated pauses: 30 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:35:35, accumulated pauses: 29 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:35:40, accumulated pauses: 29 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:35:45, accumulated pauses: 29 ms, max pause: 1 ms, pauses over 1000 ms: 0
13:35:50, accumulated pauses: 33 ms, max pause: 2 ms, pauses over 1000 ms: 0
13:35:55, accumulated pauses: 29 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:36:00, accumulated pauses: 33 ms, max pause: 4 ms, pauses over 1000 ms: 0
13:36:05, accumulated pauses: 29 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:36:10, accumulated pauses: 31 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:36:15, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:36:20, accumulated pauses: 28 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:36:25, accumulated pauses: 26 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:36:30, accumulated pauses: 26 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:36:35, accumulated pauses: 434 ms, max pause: 406 ms, pauses over 1000 ms: 0
```

Standard output for `com.hazelcast.jet.core.metrics.MetricsTest.test_sourceSinkTag`:
```
Hiccups measured while running test 'test_sourceSinkTag(com.hazelcast.jet.core.metrics.MetricsTest):'
13:30:35, accumulated pauses: 39 ms, max pause: 14 ms, pauses over 1000 ms: 0
13:30:40, accumulated pauses: 27 ms, max pause: 2 ms, pauses over 1000 ms: 0
13:30:45, accumulated pauses: 51 ms, max pause: 7 ms, pauses over 1000 ms: 0
13:30:50, accumulated pauses: 27 ms, max pause: 1 ms, pauses over 1000 ms: 0
13:30:55, accumulated pauses: 47 ms, max pause: 5 ms, pauses over 1000 ms: 0
13:31:00, accumulated pauses: 34 ms, max pause: 3 ms, pauses over 1000 ms: 0
13:31:05, accumulated pauses: 23 ms, max pause: 1 ms, pauses over 1000 ms: 0
13:31:10, accumulated pauses: 56 ms, max pause: 8 ms, pauses over 1000 ms: 0
13:31:15, accumulated pauses: 22 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:31:20, accumulated pauses: 105 ms, max pause: 83 ms, pauses over 1000 ms: 0
13:31:25, accumulated pauses: 21 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:31:30, accumulated pauses: 23 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:31:35, accumulated pauses: 22 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:31:40, accumulated pauses: 24 ms, max pause: 1 ms, pauses over 1000 ms: 0
13:31:45, accumulated pauses: 22 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:31:50, accumulated pauses: 25 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:31:55, accumulated pauses: 43 ms, max pause: 4 ms, pauses over 1000 ms: 0
13:32:00, accumulated pauses: 36 ms, max pause: 3 ms, pauses over 1000 ms: 0
13:32:05, accumulated pauses: 28 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:32:10, accumulated pauses: 44 ms, max pause: 14 ms, pauses over 1000 ms: 0
13:32:15, accumulated pauses: 29 ms, max pause: 1 ms, pauses over 1000 ms: 0
13:32:20, accumulated pauses: 29 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:32:25, accumulated pauses: 28 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:32:30, accumulated pauses: 28 ms, max pause: 0 ms, pauses over 1000 ms: 0
13:32:35, accumulated pauses: 294 ms, max pause: 281 ms, pauses over 1000 ms: 0
```

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsContext.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsContext.java
@@ -210,7 +210,7 @@ public class MetricsContext implements DynamicMetricsProvider {
 
         void threadSafeForEach(BiConsumer<? super String, ? super AbstractMetric> consumer) {
             if (threadSafeStore != null) {
-                threadLocalStore.forEach(consumer);
+                threadSafeStore.forEach(consumer);
             }
         }
     }

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsContext.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsContext.java
@@ -26,12 +26,9 @@ import com.hazelcast.jet.core.metrics.MetricTags;
 import com.hazelcast.jet.core.metrics.Unit;
 
 import javax.annotation.Nonnull;
-import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.atomic.AtomicLongFieldUpdater;
-import java.util.function.BiConsumer;
 import java.util.function.BiFunction;
 
 public class MetricsContext implements DynamicMetricsProvider {
@@ -39,7 +36,7 @@ public class MetricsContext implements DynamicMetricsProvider {
     private static final BiFunction<String, Unit, AbstractMetric> CREATE_SINGLE_WRITER_METRIC = SingleWriterMetric::new;
     private static final BiFunction<String, Unit, AbstractMetric> CREATE_THREAD_SAFE_METRICS = ThreadSafeMetric::new;
 
-    private final MetricsStore store = new MetricsStore();
+    private volatile Map<String, AbstractMetric> metrics;
 
     Metric metric(String name, Unit unit) {
         return metric(name, unit, CREATE_SINGLE_WRITER_METRIC);
@@ -50,15 +47,28 @@ public class MetricsContext implements DynamicMetricsProvider {
     }
 
     private Metric metric(String name, Unit unit, BiFunction<String, Unit, AbstractMetric> metricSupplier) {
-        return store.localGet(name, unit, metricSupplier);
+        if (metrics == null) { //first metric being stored
+            metrics = new ConcurrentHashMap<>();
+        }
+
+        AbstractMetric metric = metrics.get(name);
+        if (metric != null) {
+            return metric;
+        }
+
+        metric = metricSupplier.apply(name, unit);
+        metrics.put(name, metric);
+
+        return metric;
     }
 
     @Override
     public void provideDynamicMetrics(MetricDescriptor tagger, MetricsCollectionContext context) {
-        MetricDescriptor withUserTag = tagger.copy().withTag(MetricTags.USER, "true");
-
-        store.threadSafeForEach((name, metric) ->
-                context.collect(withUserTag, name, ProbeLevel.INFO, toProbeUnit(metric.unit()), metric.get()));
+        if (metrics != null) {
+            MetricDescriptor withUserTag = addUserTag(tagger);
+            metrics.forEach((name, metric) ->
+                    context.collect(withUserTag, name, ProbeLevel.INFO, toProbeUnit(metric.unit()), metric.get()));
+        }
     }
 
     private static MetricDescriptor addUserTag(MetricDescriptor tagger) {
@@ -174,44 +184,6 @@ public class MetricsContext implements DynamicMetricsProvider {
         @Override
         protected long get() {
             return VOLATILE_VALUE_UPDATER.get(this);
-        }
-    }
-
-    /**
-     * Map for storing metrics on a per name basis. Meant to offer fast,
-     * single-threaded access, but also a loosely consistent, thread safe
-     * way to iterate over the content.
-     */
-    private static class MetricsStore {
-
-        private Map<String, AbstractMetric> threadLocalStore;
-        private volatile ConcurrentMap<String, AbstractMetric> threadSafeStore;
-
-        /**
-         * Always called on the same thread, the one executing the Tasklet this context belongs to.
-         */
-        Metric localGet(String name, Unit unit, BiFunction<String, Unit, AbstractMetric> metricSupplier) {
-            if (threadLocalStore == null) { //first metric being stored
-                threadLocalStore = new HashMap<>();
-                threadSafeStore = new ConcurrentHashMap<>();
-            }
-
-            AbstractMetric metric = threadLocalStore.get(name);
-            if (metric != null) {
-                return metric;
-            }
-
-            metric = metricSupplier.apply(name, unit);
-            threadLocalStore.put(name, metric);
-            threadSafeStore.put(name, metric);
-
-            return metric;
-        }
-
-        void threadSafeForEach(BiConsumer<? super String, ? super AbstractMetric> consumer) {
-            if (threadSafeStore != null) {
-                threadSafeStore.forEach(consumer);
-            }
         }
     }
 

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsContext.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsContext.java
@@ -28,7 +28,10 @@ import com.hazelcast.jet.core.metrics.Unit;
 import javax.annotation.Nonnull;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import java.util.function.BiConsumer;
 import java.util.function.BiFunction;
 
 public class MetricsContext implements DynamicMetricsProvider {
@@ -36,10 +39,7 @@ public class MetricsContext implements DynamicMetricsProvider {
     private static final BiFunction<String, Unit, AbstractMetric> CREATE_SINGLE_WRITER_METRIC = SingleWriterMetric::new;
     private static final BiFunction<String, Unit, AbstractMetric> CREATE_THREAD_SAFE_METRICS = ThreadSafeMetric::new;
 
-    private String onlyName;
-    private AbstractMetric onlyMetric;
-
-    private Map<String, AbstractMetric> metrics;
+    private final MetricsStore store = new MetricsStore();
 
     Metric metric(String name, Unit unit) {
         return metric(name, unit, CREATE_SINGLE_WRITER_METRIC);
@@ -50,47 +50,15 @@ public class MetricsContext implements DynamicMetricsProvider {
     }
 
     private Metric metric(String name, Unit unit, BiFunction<String, Unit, AbstractMetric> metricSupplier) {
-        if (metrics == null) { //at most one already defined metric
-            if (onlyMetric == null) { //no already defined metrics
-                onlyMetric = metricSupplier.apply(name, unit);
-                onlyName = name;
-                return onlyMetric;
-            } else { //one single already defined metric
-                if (name.equals(onlyName)) { //single already defined metric same as the requested one
-                    return onlyMetric;
-                } else { //single already defined metric different from the requested one
-                    metrics = new HashMap<>();
-                    metrics.put(onlyName, onlyMetric);
-
-                    onlyMetric = null;
-                    onlyName = null;
-
-                    AbstractMetric metric = metricSupplier.apply(name, unit);
-                    metrics.put(name, metric);
-                    return metric;
-                }
-            }
-        } else { //multiple metrics already defined
-            AbstractMetric metric = metrics.get(name);
-            if (metric == null) { //requested metric not yet defined
-                metric = metricSupplier.apply(name, unit);
-                metrics.put(name, metric);
-            }
-            return metric;
-        }
+        return store.localGet(name, unit, metricSupplier);
     }
 
     @Override
     public void provideDynamicMetrics(MetricDescriptor tagger, MetricsCollectionContext context) {
-        if (onlyMetric != null) {
-            context.collect(
-                    addUserTag(tagger), onlyName, ProbeLevel.INFO, toProbeUnit(onlyMetric.unit()), onlyMetric.get()
-            );
-        } else if (metrics != null) {
-            MetricDescriptor withUserTag = addUserTag(tagger);
-            metrics.forEach((name, metric) ->
-                    context.collect(withUserTag, name, ProbeLevel.INFO, toProbeUnit(metric.unit()), metric.get()));
-        }
+        MetricDescriptor withUserTag = tagger.copy().withTag(MetricTags.USER, "true");
+
+        store.threadSafeForEach((name, metric) ->
+                context.collect(withUserTag, name, ProbeLevel.INFO, toProbeUnit(metric.unit()), metric.get()));
     }
 
     private static MetricDescriptor addUserTag(MetricDescriptor tagger) {
@@ -206,6 +174,44 @@ public class MetricsContext implements DynamicMetricsProvider {
         @Override
         protected long get() {
             return VOLATILE_VALUE_UPDATER.get(this);
+        }
+    }
+
+    /**
+     * Map for storing metrics on a per name basis. Meant to offer fast,
+     * single-threaded access, but also a loosely consistent, thread safe
+     * way to iterate over the content.
+     */
+    private static class MetricsStore {
+
+        private Map<String, AbstractMetric> threadLocalStore;
+        private volatile ConcurrentMap<String, AbstractMetric> threadSafeStore;
+
+        /**
+         * Always called on the same thread, the one executing the Tasklet this context belongs to.
+         */
+        Metric localGet(String name, Unit unit, BiFunction<String, Unit, AbstractMetric> metricSupplier) {
+            if (threadLocalStore == null) { //first metric being stored
+                threadLocalStore = new HashMap<>();
+                threadSafeStore = new ConcurrentHashMap<>();
+            }
+
+            AbstractMetric metric = threadLocalStore.get(name);
+            if (metric != null) {
+                return metric;
+            }
+
+            metric = metricSupplier.apply(name, unit);
+            threadLocalStore.put(name, metric);
+            threadSafeStore.put(name, metric);
+
+            return metric;
+        }
+
+        void threadSafeForEach(BiConsumer<? super String, ? super AbstractMetric> consumer) {
+            if (threadSafeStore != null) {
+                threadLocalStore.forEach(consumer);
+            }
         }
     }
 

